### PR TITLE
[model,tests] feat: add repo-native BAGEL atomic SFT primitives

### PR DESCRIPTION
**Pull request by BorisYang326**
### What does this PR do?

Part of #295. This PR consumes the generic visual-reflection contracts introduced by #313, which has now merged into `main` at `6f9fec2`.

The branch was synchronized with merged `main` in `a7f4824`, and the current review-fix commit is `07f4b97`. The final comparison against the #313 merge base contains only:

~~~text
A  tests/pipelines/test_bagel_sft_model_on_cpu.py
A  verl_omni/pipelines/bagel_flow_grpo/bagel_sft_model.py
~~~

This PR adds the repo-native BAGEL model boundary for three offline atomic SFT primitives:

- `t2i`: prompt to target-image flow-matching loss;
- `reflect`: prompt plus current image to structured Reflection / Action / Edit causal-LM loss;
- `edit`: prompt plus current image plus delta Edit to next-image flow-matching loss.

### Model and API changes

- Add `BagelForSFT` as an SFT-specific sibling of the existing `BagelForTraining` compatibility boundary.
- Keep `bagel_model.py` identical to the #313 merge base; the existing FlowGRPO implementation is no longer modified by this PR.
- Keep every shared backbone parameter registered under the original names, including `layers.N.*`, without adding wrapper prefixes or a duplicate backbone.
- Load and validate the published BAGEL language head, SigLIP vision tower, connector, position grids, and frozen VAE checkpoint tensors.
- Add fail-closed checkpoint key and shape auditing, including the published SigLIP linear-patch-weight conversion.
- Implement per-sample normalized CE and flow-matching losses for the three homogeneous primitives.
- Keep the `bagel_flow_grpo` package root narrow: it continues to expose only `BagelDiffusion` and `BagelPipelineWithLogProb`.
- Restrict `bagel_sft_model.__all__` to `BagelForSFT`, `BagelSFTConfig`, and `BagelSFTOutput`.

### SFT-only transformer boundary

All SFT-specific attention, padding, and position handling now lives in `bagel_sft_model.py` through private `_BagelSFTMoTAttention` and `_BagelSFTMoTLayer` types.

- Fully valid original T2I inputs delegate to the unchanged `BagelForTraining.forward()` path.
- Right-padded or target-masked T2I batches, reflection, and edit use the SFT-only segment-attention path.
- Transformer execution continues through `layer(...)`, preserving layer-level FSDP pre-forward, unshard, and reshard hooks.
- `BagelForSFT._no_split_modules` names the actual `_BagelSFTMoTLayer` runtime boundary.
- Checkpoint and FSDP layer keys remain compatible with the existing `layers.N.*` naming.

The compatibility regression uses original FlowGRPO inputs, compares exact forward outputs, applies the same deterministic scalar objective, and compares every shared parameter gradient and `None`-gradient state.

### Attention and padding correctness

The optional masks use `True == valid` and preserve existing all-valid behavior when omitted:

- `target_patch_mask` masks target latent supervision and target attention keys;
- `current_patch_mask` masks padded clean VAE latent tokens in the edit conditioning prefix;
- `current_vit_patch_mask` masks padded vision tokens in reflection and edit conditioning.

SigLIP valid queries cannot read invalid patch keys. Invalid SigLIP queries retain only their own diagonal to avoid fully masked SDPA rows. The same normalized vision and clean masks are propagated into the outer BAGEL MoT sequence, while image boundary markers remain valid.

Regression tests modify only masked clean latent or pixel patches and verify that valid reflection logits and edit velocities do not change.

### Scope boundary

This PR intentionally does not add:

- raw PR1 trajectory-to-tensor collation;
- an FSDP2/LoRA trainer or task sampler;
- data synthesis or conversion commands;
- a bounded reflection-loop evaluator;
- online agent/tool rollout or GRPO changes.

Those remain later RFC stages. In particular, PR3 owns the trainer and distributed normalization boundary.

### Duplicate-work check

The required searches were performed for `295 in:body`, BAGEL SFT, visual reflection, and atomic SFT.

- #313 is this PR's merged generic data-contract dependency, not a duplicate.
- #217 is a standalone Qwen-Image recipe-style T2I/edit trainer with LoRA/FSDP and CoRT row expansion. This PR instead implements repo-native BAGEL T2I/reflect/edit model forwards, checkpoint loading, MoT routing, and attention/loss masks; it does not add a trainer.
- #329 defines online agent/tool trajectories and token-level GRPO runtime behavior. This PR implements offline BAGEL atomic SFT primitives and does not modify `agent_loop`, tool calls, rollout serialization, or GRPO.

No open PR duplicates this BAGEL model-layer scope.

### Tests

PR1 dependency regression suite:

~~~text
106 passed in 0.19s
~~~

PR2 BAGEL SFT CPU suite:

~~~text
27 passed, 1 environment warning in 3.83s
~~~

The warning is the environment's existing `pynvml` deprecation warning.

The suites were run with an in-process namespace-package bootstrap because the available Python 3.10 environment does not install the repository's full `tensordict`, `verl`, and `ray` auto-registration stack. The bootstrap only bypasses top-level registration; production BAGEL and visual-reflection modules are loaded directly from this worktree.

Additional checks:

~~~text
Ruff 0.12.2 lint: passed
Ruff 0.12.2 format check: passed
Python compileall: passed
License header check: passed
Test directory structure check: passed
git diff --check: passed
bagel_model.py versus 6f9fec2: no diff
~~~

A two-GPU FSDP1 smoke test on two A100-SXM4-80GB devices also exercised transformer auto-wrap, `_BagelSFTMoTLayer`, right-padded prompts, target patch masking, the SFT-only attention path, gradient checkpointing, forward/backward, and finite local gradient shards:

~~~text
FSDP_SMOKE_OK units=3 gradients=56 loss=1.605868
~~~

### Contribution disclosure

AI assistance was used to inspect review feedback, design and implement the changes and regression tests, run verification, and update this PR. The human submitter directed the scope and architectural decisions and authorized the commit, push, PR update, and inline review reply.

### Checklist

- [x] Required duplicate-work searches completed.
- [x] PR title follows the repository format.
- [x] PR1 and PR2 targeted CPU suites pass.
- [x] Ruff, compilation, license, structure, and whitespace checks pass.
- [x] Original-input forward/backward parity is covered.
- [x] Two-GPU FSDP1 smoke test passes.
- [x] `bagel_model.py` is unchanged from the #313 merge base.

### CODE REVIEW

File: verl_omni/pipelines/bagel_flow_grpo/bagel_model.py
Comment (zhtmike) [inline verl_omni/pipelines/bagel_flow_grpo/bagel_model.py:1]: what happens here?

&

add a test to make sure the forward & backward pass unchanged for the original inputs if you want to modify this file.

cc @wtomin

Comment (BorisYang326) [reply verl_omni/pipelines/bagel_flow_grpo/bagel_model.py:1]: Moved all SFT-specific attention, padding, and position handling into `bagel_sft_model.py` in `07f4b97`. `bagel_model.py` is now identical to the #313 merge base, so the existing FlowGRPO implementation is no longer modified by this PR.

The SFT layer delegates to the original MoT forward path when no SFT segment mask is present and uses its SFT-only masked path for reflection/edit and padded T2I batches. The registered backbone parameter names remain under `layers.N.*` for checkpoint and FSDP compatibility.

I also replaced the previous forward-only comparison with an original-input forward/backward parity test that compares exact outputs and every shared parameter gradient, including matching `None`-gradient states. A two-GPU FSDP1 smoke with gradient checkpointing also completed successfully.

**Review by wtomin — APPROVED** [review]
LGTM.

File: verl_omni/pipelines/bagel_flow_grpo/bagel_model.py
Comment (zhtmike) [reply verl_omni/pipelines/bagel_flow_grpo/bagel_model.py:1]: thx
